cli/flox-watchdog/src/process/macos.rs

```diff
@@ -0,0 +1,243 @@
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::AtomicBool;
+use std::sync::Arc;
+
+use anyhow::{anyhow, bail, Context};
+use flox_rust_sdk::models::env_registry::{activation_pids, ActivationPid};
+use kqueue::Ident;
+use tracing::warn;
+
+use super::{Error, WaitResult, Watcher, WATCHER_SLEEP_INTERVAL};
+
+/// Stores a list of PIDs to monitor via kqueue and waits for them to terminate.
+#[derive(Debug)]
+pub struct MacOsWatcher {
+    pub original_pid: ActivationPid,
+    pub pids_watching: HashSet<ActivationPid>,
+    pub reg_path: PathBuf,
+    pub hash: String,
+    pub should_terminate_flag: Arc<AtomicBool>,
+    pub should_clean_up_flag: Arc<AtomicBool>,
+    pub kqueue_watcher: Option<kqueue::Watcher>,
+}
+
+impl MacOsWatcher {
+    pub fn new(
+        pid: ActivationPid,
+        reg_path: impl AsRef<Path>,
+        hash: impl AsRef<str>,
+        should_terminate_flag: Arc<AtomicBool>,
+        should_clean_up_flag: Arc<AtomicBool>,
+    ) -> Self {
+        MacOsWatcher {
+            original_pid: pid,
+            pids_watching: HashSet::new(),
+            reg_path: PathBuf::from(reg_path.as_ref()),
+            hash: String::from(hash.as_ref()),
+            should_terminate_flag,
+            should_clean_up_flag,
+            kqueue_watcher: None,
+        }
+    }
+
+    /// Initialize the internal kqueue watcher and add the PID for the activation
+    /// that spawned this watchdog.
+    fn init_kqueue(&mut self) -> Result<(), Error> {
+        let mut kw = kqueue::Watcher::new().context("failed to create kqueue watcher")?;
```

Comment:
Doesn't this regress https://github.com/flox/flox/issues/2002 and https://github.com/flox/flox/pull/2018 by leaving services running if the `wait_for_termination` starts after a single `flox activate -s` has exited, the error bubbles through `wait_for_termination` and `main`, and we exit without calling `cleanup`?

I would have expected the integration tests to catch it but we also have trust issues with those tests 🙈

Reply:
Not that line, but line 181 in `main.rs` which just bails on an error. I don't think there's a way to avoid starting the watchdog after its parent has terminated, and thus there's no way to prevent that error, but yes this regresses whether we clean up.

Good catch though, I was leaning on integration tests for any behavior that I missed while I was out at RustConf and with COVID.

Reply:
Also, the error wouldn't be when a single activation has exited since in order for this watchdog to have created a watcher in the first place we would have already checked that there are no other watchdogs and thus no other activation PIDs in the registry for it to watch.

Reply:
By "single activation" I was thinking of something short-lived like `flox activate -s -- redis-cli ping` where in no guaranteed order:
- services are started
- watchdog is started
- redis-cli is started
- redis-cli exits
- maybe watchdog started monitoring the redis-cli PID
- maybe services are shutdown
- watchdog exits

Reply:
Having a bit of a hard time parsing this one. Are you saying that this isn't covered by my explanation above, or are you just clarifying your original comment?

Reply:
Addressed the regression in https://github.com/flox/flox/pull/2192/commits/bd9590660980fa92769477378385330a2c3bdac4

Reply:
I was just clarifying because it sound from "the error wouldn't be when a single activation has exited" like we weren't talking about the same thing.